lib/quicly.c

```diff
@@ -4891,13 +4939,24 @@ int quicly_receive(quicly_conn_t *conn, struct sockaddr *dest_addr, struct socka
     conn->super.stats.num_bytes.received += packet->octets.len;
 
     /* state updates, that are triggered by the receipt of a packet */
-    if (epoch == QUICLY_EPOCH_HANDSHAKE && conn->initial != NULL) {
+    switch (epoch) {
+    case QUICLY_EPOCH_INITIAL:
+        /* update max_ingress_udp_payload_size if necessary */
+        if (conn->initial->max_ingress_udp_payload_size < packet->datagram_size)
+            conn->initial->max_ingress_udp_payload_size = packet->datagram_size;
+        break;
```

Comment:
This applies to the client as well, right? So if the server sends a larger Initial than the client, then the client also ratchets up its own MTU?

Reply:
I'm not sure if we should do that.

When the first flights from either endpoints do not reach the peers, the client has the chance to reconnect, possibly using a different protocol (e.g., TCP).

But once the first round-trip is complete, the damage of the connection not moving forward becomes bigger, as it would likely end up as an error in the application.

Reply:
Ok, fair enough...  we should not do this without having some protection around it. So, as it stands, how does this code only apply at the server? Doesn't quicly_receive get called at both endpoints, and the server's first Initial might be larger than the client's udp_payload_size, right?

Reply:
> So, as it stands, how does this code only apply at the server?

This value is called "max_ingress_udp_payload_size", it records the size of the largest datagram that the endpoint received. I will rename the field to "largest_ingress_udp_payload_size", as I believe we use "largest" for variables retaining the observed values.

The maximum size of the egress datagrams are determined in the `server_collected_extensions` function (which is only called on the server-side, as the name implies), taking this `max_ingress_udp_payload_size` and TP.max_udp_payload_size of both endpoints into consideration.